File: src/main/java/br/com/ingenieux/mojo/jbake/util/DirWatcher.java
package br.com.ingenieux.mojo.jbake.util;

import org.apache.commons.io.monitor.FileAlterationListenerAdaptor;
import org.apache.commons.io.monitor.FileAlterationMonitor;
import org.apache.commons.io.monitor.FileAlterationObserver;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

/**
 * Example to watch a directory (or tree) for changes to files.
 */

public class DirWatcher {

  private final FileAlterationObserver observer;

  private final FileAlterationMonitor monitor;

  private final BlockingQueue<Long> changeQueue = new ArrayBlockingQueue<Long>(1);

  /**
   * Creates a WatchService and registers the given directory
   */
  public DirWatcher(File dir) throws IOException {
    this.observer = new FileAlterationObserver(dir);
    this.monitor = new FileAlterationMonitor(1000, observer);

    observer.addListener(new FileAlterationListenerAdaptor() {
      @Override
      public void onFileCreate(File file) {
        onUpdated();
      }

      @Override
      public void onFileChange(File file) {
        onUpdated();
      }
    });
  }

  public void start() throws Exception {
    monitor.start();
  }

  public void stop() {
    try {
      monitor.stop();
    } catch (Exception exc) {
    }
  }

  private void onUpdated() {
    changeQueue.add(Long.valueOf(System.currentTimeMillis()));
  }

  /**
   * Process all events for keys queued to the watcher
   */
  public Long processEvents() throws InterruptedException {
    return changeQueue.poll(1, TimeUnit.SECONDS);
  }
}

File: jbake-maven-plugin/src/main/java/org/jbake/maven/util/DirWatcher.kt
package org.jbake.maven.util

import org.apache.commons.io.monitor.FileAlterationListenerAdaptor
import org.apache.commons.io.monitor.FileAlterationMonitor
import org.apache.commons.io.monitor.FileAlterationObserver
import java.io.File
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.TimeUnit

/**
 * Watches a directory for file changes and queues modification timestamps.
 */
class DirWatcher(dir: File) {

    private val changeQueue = ArrayBlockingQueue<Long>(1)

    private val monitor = FileAlterationMonitor(1000, FileAlterationObserver(dir).apply {
        addListener(object : FileAlterationListenerAdaptor() {
            override fun onFileCreate(file: File) = recordChange()
            override fun onFileChange(file: File) = recordChange()
        })
    })

    fun start() = monitor.start()

    fun stop() = runCatching { monitor.stop() }

    private fun recordChange() {
        try {
            changeQueue.put(System.currentTimeMillis())
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
        }
    }

    /** Polls for file change events, returning the timestamp of the last change or null if no changes. */
    fun processEvents(): Long? = changeQueue.poll(1, TimeUnit.SECONDS)
}
